src/fetchInterceptor.js

```diff
@@ -0,0 +1,249 @@
+      shouldFetch: true,
+      accessToken: null,
+    }
+  }
+
+  shouldIntercept(requestUnit) {
+    const { request } = requestUnit;
+
+    return Promise.all([requestUnit, this.config.shouldIntercept(request)])
+      .then(([requestUnit, shouldIntercept]) =>
+        ({ ...requestUnit, shouldIntercept })
+      );
+  }
+
+  shouldFetch(requestUnit) {
+    const { request } = requestUnit;
+
+    if (this.config.shouldFetch) {
+      return Promise.all([requestUnit, this.config.shouldFetch(request)])
+        .then(([requestUnit, shouldFetch]) =>
+          ({ ...requestUnit, shouldFetch })
+        );
+    }
+
+    return requestUnit;
+  }
+
+  authorizeRequest(requestUnit) {
+    const { shouldIntercept } = requestUnit;
+
+    if (shouldIntercept) {
+      const { request } = requestUnit;
+      const { accessToken } = this.accessTokenProvider.getAuthorization();
+
+      if (request && accessToken){
+        return Promise.all([
+          requestUnit,
+          accessToken,
+          this.config.authorizeRequest(request, accessToken),
+        ]).then(([requestUnit, accessToken, request]) =>
+          ({ ...requestUnit, accessToken, request })
+        );
+      }
+    }
+
+    return requestUnit;
+  }
+
+  fetchRequest(requestUnit) {
+    const { shouldFetch } = requestUnit;
+
+    if (shouldFetch) {
+      const { request } = requestUnit;
+      const { fetch } = this;
+      return Promise.all([requestUnit, fetch(request)])
+        .then(([requestUnit, response]) =>
+          ({...requestUnit, response})
+        );
+    }
+
+    return requestUnit;
+  }
+
+  shouldInvalidateAccessToken(requestUnit) {
+    const { shouldIntercept } = requestUnit;
+
+    if (shouldIntercept && this.config.shouldInvalidateAccessToken) {
+      const { response } = requestUnit;
+      // check if response invalidates access token
+      return Promise.all([
+        requestUnit,
+        this.config.shouldInvalidateAccessToken(response),
+      ]).then(([requestUnit, shouldInvalidateAccessToken]) =>
+        ({ ...requestUnit, shouldInvalidateAccessToken })
+      );
+    }
+
+    return requestUnit;
+  }
+
+  invalidateAccessToken(requestUnit) {
+    const { shouldIntercept, shouldInvalidateAccessToken } = requestUnit;
+
+    if (shouldIntercept && shouldInvalidateAccessToken) {
+      this.accessTokenProvider.refresh();
+    }
+
+    return requestUnit;
+  }
+
+  handleUnauthorizedRequest(requestUnit) {
+    const { shouldIntercept } = requestUnit;
+
+    if (shouldIntercept) {
+      const { response } = requestUnit;
+
+      // we only care for unauthorized responses
+      if (isResponseUnauthorized(response)) {
+        return Promise.all([requestUnit, this.accessTokenProvider.refresh()])
+          .then(([requestUnit, accessToken]) => ({
+            ...requestUnit,
+            accessToken,
+            shouldFetch: !!accessToken,
+          }))
+          .then(this.authorizeRequest)
+          .then(this.fetchRequest)
```

Comment:
This block  in `if (isResponseUnauthorized(response)) {` is problematic, what happens if it fails twice? You only have one retry right? Should you  reuse fetchWithRetry and move request through flow once again?